[Bug]: Module not found: Error: Can't resolve `@storybook/addon-docs/mdx-react-shim`

### Describe the bug

When processing an `.mdx` file in my repo, storybook produces the following error:

```
ERR! => Failed to build the preview
ERR! Module not found: Error: Can't resolve '@storybook/addon-essentials/docs/mdx-react-shim' in '/tmp/test/packages/a/src/Foo'
ERR! ModuleNotFoundError: Module not found: Error: Can't resolve '@storybook/addon-essentials/docs/mdx-react-shim' in '/tmp/test/packages/a/src/Foo'
```

This is similar to #21716, but the solutions there either do not work or do not seem very satisfactory to me.

### To Reproduce

1. Extract [test.zip](https://github.com/storybookjs/storybook/files/11873776/test.zip)
2. In that repo, using pnpm 8, run `pnpm install && pnpm --dir=packages/storybook run build`


### System

```shell
Environment Info:

  System:
    OS: Linux 6.3 Debian GNU/Linux trixie/sid
    CPU: (8) x64 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
  Binaries:
    Node: 18.13.0 - /usr/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 9.2.0 - /usr/bin/npm

Also of note

    pnpm: 8.6.3 - /usr/local/bin/pnpm
```


### Additional context

The problem in the test repo seems to be at https://github.com/storybookjs/storybook/blob/f57e67f302865930f1bca01a977d0624ea663f19/code/addons/essentials/src/docs/preset.ts#L6 where it assumes that the code passed that option will have `@storybook/addon-essentials` available, which may not be the case in pnpm with hoisting disabled or in yarn when using PnP. Passing it through `require.resolve()` (or something similar to get an absolute path for `@storybook/addon-essentials/docs/mdx-react-shim`) should work there.

If `@storybook/addon-docs` is being used instead, code at https://github.com/storybookjs/storybook/blob/f57e67f302865930f1bca01a977d0624ea663f19/code/addons/docs/src/preset.ts#L59 makes a similar assumption, and a similar solution should work there.

I also see code in `@storybook/builder-vite` at https://github.com/storybookjs/storybook/blob/f57e67f302865930f1bca01a977d0624ea663f19/code/builders/builder-vite/src/plugins/mdx-plugin.ts#L36 that assumes that `@storybook/addon-docs` will be available when the option is being resolved. In this case a simple `require.resolve()` won't always work since there is no dependency or peer dependency of `@storybook/builder-vite` on `@storybook/addon-docs` to make it available to it.

In light of the above, here are some unsatisfactory workarounds for the problem:

* Re-enable pnpm's indirect-dep fallback hoisting, which exists to sometimes work around broken deps like this.
* Add an otherwise-unnecessary dep on `@storybook/addon-essentials` or `@storybook/addon-docs` (depending on what exactly is in `main.cjs`) to either the root `package.json` or to `projects/a/package.json` so the package will be available in the context where the option is being resolved.
* Add the following additional addon in `.storybook/main.cjs` to override the problematic option with a usable value:
   ```
        {
            name: '@storybook/addon-docs',
            options: {
                mdxPluginOptions: {
                    mdxCompileOptions: {
                        providerImportSource: require.resolve( '@storybook/addon-docs/mdx-react-shim' ),
                    },
                },
            },
        },
   ```
* Use `NODE_PATH` when running storybook to have it always search `packages/storybook/node_modules/` for dependencies in addition to the normal locations.

--- a/.gitignore
+++ b/.gitignore
@@ -50,3 +50,5 @@ code/playwright-results/
 code/playwright-report/
 code/playwright/.cache/
 code/bench-results/
+
+/packs

--- a/code/addons/docs/src/preset.ts
+++ b/code/addons/docs/src/preset.ts
@@ -50,7 +50,7 @@ async function webpack(
     skipCsf: true,
     ...mdxPluginOptions,
     mdxCompileOptions: {
-      providerImportSource: '@storybook/addon-docs/mdx-react-shim',
+      providerImportSource: require.resolve('@storybook/addon-docs/dist/shims/mdx-react-shim'),
       ...mdxPluginOptions.mdxCompileOptions,
       remarkPlugins: [remarkSlug, remarkExternalLinks].concat(
         mdxPluginOptions?.mdxCompileOptions?.remarkPlugins ?? []

--- a/code/addons/docs/src/preset.ts
+++ b/code/addons/docs/src/preset.ts
@@ -1,4 +1,5 @@
 import fs from 'fs-extra';
+import { dirname, join } from 'path';
 import remarkSlug from 'remark-slug';
 import remarkExternalLinks from 'remark-external-links';
 import { dedent } from 'ts-dedent';
@@ -50,7 +51,10 @@ async function webpack(
     skipCsf: true,
     ...mdxPluginOptions,
     mdxCompileOptions: {
-      providerImportSource: require.resolve('@storybook/addon-docs/dist/shims/mdx-react-shim'),
+      providerImportSource: join(
+        dirname(require.resolve('@storybook/addon-docs/package.json')),
+        '/dist/shims/mdx-react-shim'
+      ),
       ...mdxPluginOptions.mdxCompileOptions,
       remarkPlugins: [remarkSlug, remarkExternalLinks].concat(
         mdxPluginOptions?.mdxCompileOptions?.remarkPlugins ?? []

--- a/code/addons/essentials/src/docs/preset.ts
+++ b/code/addons/essentials/src/docs/preset.ts
@@ -3,6 +3,8 @@ export * from '@storybook/addon-docs/dist/preset';
 
 export const mdxLoaderOptions = async (config: any) => {
   // eslint-disable-next-line no-param-reassign
-  config.mdxCompileOptions.providerImportSource = '@storybook/addon-essentials/docs/mdx-react-shim';
+  config.mdxCompileOptions.providerImportSource = require.resolve(
+    '@storybook/addon-docs/dist/shims/mdx-react-shim'
+  );
   return config;
 };

--- a/code/addons/essentials/src/docs/preset.ts
+++ b/code/addons/essentials/src/docs/preset.ts
@@ -1,10 +1,13 @@
-/* eslint-disable import/export */
+import { dirname, join } from 'path';
+
+// eslint-disable-next-line import/export
 export * from '@storybook/addon-docs/dist/preset';
 
 export const mdxLoaderOptions = async (config: any) => {
   // eslint-disable-next-line no-param-reassign
-  config.mdxCompileOptions.providerImportSource = require.resolve(
-    '@storybook/addon-docs/dist/shims/mdx-react-shim'
+  config.mdxCompileOptions.providerImportSource = join(
+    dirname(require.resolve('@storybook/addon-docs/package.json')),
+    '/dist/shims/mdx-react-shim'
   );
   return config;
 };

--- a/code/builders/builder-vite/src/plugins/mdx-plugin.ts
+++ b/code/builders/builder-vite/src/plugins/mdx-plugin.ts
@@ -3,6 +3,7 @@ import type { Plugin } from 'vite';
 import remarkSlug from 'remark-slug';
 import remarkExternalLinks from 'remark-external-links';
 import { createFilter } from 'vite';
+import { dirname, join } from 'path';
 
 const isStorybookMdx = (id: string) => id.endsWith('stories.mdx') || id.endsWith('story.mdx');
 
@@ -33,7 +34,10 @@ export async function mdxPlugin(options: Options): Promise<Plugin> {
       const mdxLoaderOptions = await options.presets.apply('mdxLoaderOptions', {
         ...mdxPluginOptions,
         mdxCompileOptions: {
-          providerImportSource: require.resolve('@storybook/addon-docs/mdx-react-shim'),
+          providerImportSource: join(
+            dirname(require.resolve('@storybook/addon-docs/package.json')),
+            '/dist/shims/mdx-react-shim'
+          ),
           ...mdxPluginOptions?.mdxCompileOptions,
           remarkPlugins: [remarkSlug, remarkExternalLinks].concat(
             mdxPluginOptions?.mdxCompileOptions?.remarkPlugins ?? []

--- a/code/builders/builder-vite/src/plugins/mdx-plugin.ts
+++ b/code/builders/builder-vite/src/plugins/mdx-plugin.ts
@@ -33,7 +33,7 @@ export async function mdxPlugin(options: Options): Promise<Plugin> {
       const mdxLoaderOptions = await options.presets.apply('mdxLoaderOptions', {
         ...mdxPluginOptions,
         mdxCompileOptions: {
-          providerImportSource: '@storybook/addon-docs/mdx-react-shim',
+          providerImportSource: require.resolve('@storybook/addon-docs/mdx-react-shim'),
           ...mdxPluginOptions?.mdxCompileOptions,
           remarkPlugins: [remarkSlug, remarkExternalLinks].concat(
             mdxPluginOptions?.mdxCompileOptions?.remarkPlugins ?? []

--- a/code/lib/preview/src/globals/runtime.ts
+++ b/code/lib/preview/src/globals/runtime.ts
@@ -4,6 +4,7 @@ import * as CHANNELS from '@storybook/channels';
 import * as CLIENT_LOGGER from '@storybook/client-logger';
 import * as CORE_EVENTS from '@storybook/core-events';
 import * as PREVIEW_API from '@storybook/preview-api';
+import * as GLOBAL from '@storybook/global';
 
 // DEPRECATED, remove in 8.0
 import * as ADDONS from '@storybook/preview-api/dist/addons';
@@ -22,6 +23,7 @@ export const values: Required<Record<keyof typeof globals, any>> = {
   '@storybook/client-logger': CLIENT_LOGGER,
   '@storybook/core-events': CORE_EVENTS,
   '@storybook/preview-api': PREVIEW_API,
+  '@storybook/global': GLOBAL,
 
   // DEPRECATED, remove in 8.0
   '@storybook/addons': ADDONS,

--- a/code/lib/preview/src/globals/types.ts
+++ b/code/lib/preview/src/globals/types.ts
@@ -1,6 +1,7 @@
 // Here we map the name of a module to their NAME in the global scope.
 export const globals = {
   '@storybook/addons': '__STORYBOOK_MODULE_ADDONS__',
+  '@storybook/global': '__STORYBOOK_MODULE_GLOBAL__',
   '@storybook/channel-postmessage': '__STORYBOOK_MODULE_CHANNEL_POSTMESSAGE__', // @deprecated: remove in 8.0
   '@storybook/channel-websocket': '__STORYBOOK_MODULE_CHANNEL_WEBSOCKET__', // @deprecated: remove in 8.0
   '@storybook/channels': '__STORYBOOK_MODULE_CHANNELS__',

--- a/scripts/run-registry.ts
+++ b/scripts/run-registry.ts
@@ -7,6 +7,8 @@ import program from 'commander';
 import { runServer, parseConfigFile } from 'verdaccio';
 import pLimit from 'p-limit';
 import type { Server } from 'http';
+import { mkdir } from 'fs/promises';
+import { PACKS_DIRECTORY } from './utils/constants';
 // @ts-expect-error (Converted from ts-ignore)
 import { maxConcurrentTasks } from './utils/concurrency';
 import { listOfPackages } from './utils/list-packages';
@@ -53,11 +55,26 @@ const currentVersion = async () => {
   return version;
 };
 
-const publish = (packages: { name: string; location: string }[], url: string) => {
+const publish = async (packages: { name: string; location: string }[], url: string) => {
   logger.log(`Publishing packages with a concurrency of ${maxConcurrentTasks}`);
 
   const limit = pLimit(maxConcurrentTasks);
   let i = 0;
+
+  /**
+   * We need to "pack" our packages before publishing to npm because our package.json files contain yarn specific version "ranges".
+   * such as "workspace:*"
+   *
+   * We can't publish to npm if the package.json contains these ranges. So with `yarn pack` we create a tarball that we can publish.
+   *
+   * However this bug exists in NPM: https://github.com/npm/cli/issues/4533!
+   * Which causes the NPM CLI to disregard the tarball CLI argument and instead re-create a tarball.
+   * But NPM doesn't replace the yarn version ranges.
+   *
+   * So we create the tarball ourselves and move it to another location on the FS.
+   * Then we change-directory to that directory and publish the tarball from there.
+   */
+  await mkdir(PACKS_DIRECTORY, { recursive: true }).catch(() => {});
 
   return Promise.all(
     packages.map(({ name, location }) =>
@@ -70,7 +87,9 @@ const publish = (packages: { name: string; location: string }[], url: string) =>
                 '.'
               )})`
             );
-            const command = `cd ${location} && yarn pack && npm publish ./package.tgz --registry ${url} --force --access restricted --ignore-scripts && rm ./package.tgz`;
+
+            const tarballFilename = `${name.replace('@', '').replace('/', '-')}.tgz`;
+            const command = `cd ${location} && yarn pack --out=${PACKS_DIRECTORY}/${tarballFilename} && cd ${PACKS_DIRECTORY} && npm publish ./${tarballFilename} --registry ${url} --force --access restricted --ignore-scripts`;
             exec(command, (e) => {
               if (e) {
                 rej(e);

--- a/scripts/utils/constants.ts
+++ b/scripts/utils/constants.ts
@@ -3,7 +3,9 @@ import { join } from 'path';
 export const AFTER_DIR_NAME = 'after-storybook';
 export const BEFORE_DIR_NAME = 'before-storybook';
 
+export const ROOT_DIRECTORY = join(__dirname, '..', '..');
 export const CODE_DIRECTORY = join(__dirname, '..', '..', 'code');
+export const PACKS_DIRECTORY = join(__dirname, '..', '..', 'packs');
 export const REPROS_DIRECTORY = join(__dirname, '..', '..', 'repros');
 export const SANDBOX_DIRECTORY = join(__dirname, '..', '..', 'sandbox');
 export const JUNIT_DIRECTORY = join(__dirname, '..', '..', 'test-results');

